### DISCUSSION TITLE
MLFlowLogger implementation: checkpoints downloaded during mlflow.pytorch.load_model.

### DISCUSSION THREAD

**Discussion by croth1-liveeo** [Lightning-AI/pytorch-lightning#19394] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
**TL;DR**: It appears to me that lightnings MLFlow logger implementation's choice to put checkpoints into mlflow's model directory might lead to excessive downloads during model loading.

# Details
It is my understanding that checkpoints are generally much larger than the models themselves, probably due to extra information such as optimizer state, etc.

## Lightning's MLflow logger implementation
Furthermore, it seems that the default behavior of lightnings mlflowlogger is to upload checkpoints at the end of the run according to the ModelCheckpoints config. The checkpoint path seems hard-coded to the `checkpoints` subdirectory of the MLflow model.

![image](https://github.com/Lightning-AI/pytorch-lightning/assets/125550214/163a821a-bc7c-4eee-b11f-015ad59f47d7)

https://github.com/Lightning-AI/pytorch-lightning/blob/8646515caca7b7281b92248f392c44aa45688a60/src/lightning/pytorch/loggers/mlflow.py#L356

## MLflows model loading behavior
the MLflow model loading code [seems to download the full models directory](https://github.com/mlflow/mlflow/blob/bf141c74ffb816e4bfce54177eef704c25849d0b/mlflow/pytorch/__init__.py#L687) when loading from a remote target.

To me this seems to mean that whenever one attempts to load such a model with `mlflow.pytorch.load_model(f"runs:/{run_id}/model")` logged with lightnings mlflow-logger implementation, one would always also download all the checkpoints. However it is my suspicion that the weights needed during unpickling are in `data/model.pth`.

# Moving the checkpoints out of the model folder?
Ultimately my question is: is there a reason for putting the checkpoints as a sub-directory into `model` directory, or would it be better to put them under a top-level `checkpoints` directory which to my understanding should avoid downloading all the checkpoints during model load?

Am I missing something here?

**Comment by croth1-liveeo** [comment]
Seems to have been us using the `model` directory for logging the model explicitly, clashing with lightnings choice of using `model/checkpoints` as their preferred checkpoint location.

Bit of a gotcha I was not foreseeing without looking deep into the code. Is there a reason for choosing `model/checkpoints` instead of letting the user set the artifact location?

**Comment by benglewis** [comment]
It would be nice if the release notes for v2.5 would mention that this behavior was changed now and there is still no configurable prefix to the checkpoint, but instead it now no longer adds `model/checkpoints`. This is a **breaking** change and this broke things in our code. I will also see if I can open a PR with the option to add a prefix

**Reply by benglewis** [reply]
I've opened an issue for this: https://github.com/Lightning-AI/pytorch-lightning/issues/20540
